wen-restart/src/wen_restart.rs

```diff
@@ -0,0 +1,152 @@
+//! The `wen-restart` module handles automatically repair in cluster restart
+
+use {
+    crate::solana::wen_restart_proto,
+    log::*,
+    prost::Message,
+    solana_gossip::cluster_info::ClusterInfo,
+    solana_ledger::blockstore::Blockstore,
+    solana_vote_program::vote_state::VoteTransaction,
+    std::{
+        fs::File,
+        io::{Error, Write},
+        path::PathBuf,
+        sync::Arc,
+    },
+};
+
+// The number of ancestor slots sent is hard coded at 81000, because that's
+// 400ms * 81000 = 9 hours, we assume most restart decisions to be made in 9
+// hours.
+const MAX_SLOTS_ON_VOTED_FORKS: u32 = 81000;
+
+pub fn wait_for_wen_restart(
+    wen_restart_path: &PathBuf,
+    last_vote: VoteTransaction,
+    blockstore: Arc<Blockstore>,
+    cluster_info: Arc<ClusterInfo>,
+) -> Result<(), Box<dyn std::error::Error>> {
+    // repair and restart option does not work without last voted slot.
+    let last_vote_slot = last_vote
+        .last_voted_slot()
+        .expect("wen_restart doesn't work if local tower is wiped");
+    let mut last_vote_fork = vec![last_vote_slot];
+    let mut slot = last_vote_slot;
+    for _ in 0..MAX_SLOTS_ON_VOTED_FORKS {
+        match blockstore.meta(slot) {
+            Ok(Some(slot_meta)) => {
+                match slot_meta.parent_slot {
+                    Some(parent_slot) => {
+                        last_vote_fork.push(parent_slot);
+                        slot = parent_slot;
+                    }
+                    None => break,
+                };
+            }
+            _ => break,
+        }
+    }
+    info!(
+        "wen_restart last voted fork {} {:?}",
+        last_vote_slot, last_vote_fork
+    );
+    last_vote_fork.reverse();
+    // Todo(wen): add the following back in after Gossip code is checked in.
+    //    cluster_info.push_last_voted_fork_slots(&last_voted_fork, last_vote.hash());
+    // The rest of the protocol will be in another PR.
+    let cur_progress = wen_restart_proto::WenRestartProgress {
+        state: wen_restart_proto::wen_restart_progress::State::Init.into(),
+        init_record: Some(wen_restart_proto::wen_restart_progress::InitRecord {
```

Comment:
```suggestion
    let current_progress = WenRestartProgress {
        state: RestartState::Init.into(),
        init_record: Some(InitRecord {
```

This would be a bit easier to read if you import the types in your use statement above. If `State` seems too generic, could alias it. Something like:
```
use {
    crate::solana::wen_restart_proto::{InitRecord, State as RestartState, WenRestartProgress},
    ...
}
```

Also, please spell out "current"